### PR TITLE
Map extension support with regex extension as first extension

### PR DESCRIPTION
**Pull request by glenna-m**

### CODE REVIEW

File: extensions/shex-map/lib/extension-utils.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  var idx = vars.indexOf(expandedVarName);
+  if (idx !== -1) 
+    throw Error("unable to process prefixes in " + expandedVarName);
+
+  vars.push(expandedVarName);
+  return expandedVarName;
+}
+
+// Unescape the backslash characters in a string (e.g., in a URL)
+function unescapeMetaChars(string) {
```
Comment (dbooth-boston) [inline extensions/shex-map/lib/extension-utils.js]: Can this use one of the standard library functions instead, such as decodeURI or decodeURIComponent?

File: extensions/shex-map/lib/extension-utils.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    expandedVarName = prefixes[onURIPattern[2]] + onURIPattern[3];
+
+  else
+    // unknown prefix
+    throw Error("Unknown prefix " + onURIPattern[2] + " in \"" + varName + "\"!");
+
+  return expandedVarName;
+}
+
+function getVariable(p1, vars, prefixes) {
```
Comment (dbooth-boston) [inline extensions/shex-map/lib/extension-utils.js]: A comment would be helpful here, saying that getVariable also writes the expanded variable into vars.  And a more mnemonic name than p1 would help too.  How about:
```
function expandVariable(var, expandedVars, prefixes) {
```

File: extensions/shex-map/lib/extensions.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    var startArgs = mapDirective.indexOf('(', 0);
+    var endArgs = mapDirective.lastIndexOf(')');
+    if (startArgs < 2 || endArgs < 4 || endArgs <= startArgs+1 || endArgs != mapDirective.length-1) 
+        throw Error("Invalid extension function: " + mapDirective + "!");
+
+    return { name:  mapDirective.substring(0, startArgs),
+             args:  mapDirective.substring(startArgs+1, endArgs) }; 
+}
+
+function lift(mapDirective, input, prefixes) {
```
Comment (dbooth-boston) [inline extensions/shex-map/lib/extensions.js]: We should check with Eric: could we standardize on calling this "raise" and "lower", to be distinct from "lift" and "drop"?   I realized that lift and drop are going to/from RDF, whereas these operations are going to/from variable bindings.

Comment (dbooth-boston) [inline extensions/shex-map/lib/extensions.js]: P.S. If Eric agrees to this naming convention, it might be easiest to do it in a separate PR.

File: extensions/shex-map/lib/extension-utils.js (hunk 2 above)
Comment (glenna-m) [inline extensions/shex-map/lib/extension-utils.js]: Agree.  This API was originally Eric's.  I actually don't like the side-effect of updating expandedVars as well as returning the individual value, but I didn't want to make more extensive changes than I already did.    I'll make this change.

File: extensions/shex-map/lib/extensions.js (hunk 3 above)
Comment (glenna-m) [inline extensions/shex-map/lib/extensions.js]: I don't care what we call it.  In his earlier references that I built this from, he called it lift/lower.  Eric, can you weigh in on this?

File: extensions/shex-map/lib/extension-utils.js (hunk 1 above)
Comment (glenna-m) [inline extensions/shex-map/lib/extension-utils.js]: No, I can't.  It would not do the same thing.   It is dealing with regex escapes, not URI escapes.

Comment (dbooth-boston) [inline extensions/shex-map/lib/extension-utils.js]: ok, thanks

Comment (dbooth-boston) [inline extensions/shex-map/lib/extension-utils.js]: ok, thanks
